packages/celotool/src/e2e-tests/validator_handshake_tests.ts

```diff
@@ -0,0 +1,263 @@
+import { eqAddress } from '@celo/utils/lib/address'
+import { assert } from 'chai'
+import _ from 'lodash'
+import sleep from 'sleep-promise'
+import Web3 from 'web3'
+import { Admin } from 'web3-eth-admin'
+import { GethRunConfig } from '../lib/interfaces/geth-run-config'
+import { GethInstanceConfig } from '../lib/interfaces/geth-instance-config'
+import { configureInstances, getContext, jsonRpc, killInstance, restartInstance } from './utils'
+
+const VALIDATORS = 3
+const PROXIED_VALIDATORS = 1
+const EPOCH = 20
+
+const TMP_PATH = '/tmp/e2e'
+
+describe('Validator handshake tests', () => {
+  const gethConfig: GethRunConfig = {
+    networkId: 1101,
+    network: 'local',
+    runPath: TMP_PATH,
+    migrate: false,
+    instances: getInstances(),
+    genesisConfig: {
+      epoch: EPOCH,
+    },
+  }
+  const context: any = getContext(gethConfig)
+
+  before(async function(this: any) {
+    this.timeout(0)
+    await context.hooks.before()
+  })
+
+  after(context.hooks.after)
+
+  describe('Validator handshake', () => {
+    before(async function() {
+      // reset the instances for each test
+      gethConfig.instances = getInstances()
+      configureInstances(gethConfig)
+      this.timeout(0)
+      await context.hooks.restart()
+    })
+
+    it('allows a validator with a new enode to identify itself to a newly peered validator whose maxpeers is 0', async function(this: any) {
+      this.timeout(20000) // 20 seconds
+      // If a consensus round fails during this test, the results are inconclusive.
+      // Retry up to two times to mitigate this issue. Restarting the nodes is not needed.
+      this.retries(2)
+
+      const val0Provider = `http://localhost:${gethConfig.instances[0].rpcport}`
+      const val0Web3 = new Web3(val0Provider)
+      const val0Admin = new Admin(val0Provider)
+      const val0InfoBefore = await val0Admin.getNodeInfo()
+      const val0EnodeBefore = val0InfoBefore.enode
+
+      const val1Provider = `http://localhost:${gethConfig.instances[1].rpcport}`
```

Comment:
The fact that this code works to get the correct node seems to be a consequence of the exact implementation of `getInstances` and the constant chosen above. Please refactor this to be less fragile in case someone comes in and makes changes to those because it would not be clear what the issue is.